nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py

```diff
@@ -440,15 +551,252 @@ def build_train_valid_test_datasets(self):
+        if self.multilingual:
+            if not (is_src_listconfig and is_tgt_listconfig):
+                raise ValueError(
+                    f"Multilingual datasets must be configured with a ListConfig for both src_file_name and tgt_file_name"
+                )
+        if is_src_listconfig and not is_tgt_listconfig or is_tgt_listconfig and not is_src_listconfig:
+            raise ValueError(
+                f"Datasets must be configured with a ListConfig for both src_file_name and tgt_file_name or neither. Found only one of them as listconfig."
+            )
+
+        if is_src_listconfig and is_tgt_listconfig:
+            if len(cfg.src_file_name) != len(cfg.tgt_file_name):
+                raise ValueError(f"Datasets must have the same number of files in src_file_name and tgt_file_name")
+
+            if cfg.concat_sampling_probabilities is None or not isinstance(
+                cfg.concat_sampling_probabilities, ListConfig
+            ):
+                raise ValueError(
+                    f"concat_sampling_probabilities must be a ListConfig with the same number of files in src_file_name and tgt_file_name, found {cfg.concat_sampling_probabilities}"
+                )
+            if len(cfg.concat_sampling_probabilities) != len(cfg.src_file_name):
+                raise ValueError(
+                    f"concat_sampling_probabilities must be of the same size as src_file_name and tgt_file_name. Provided size {len(cfg.concat_sampling_probabilities)}, number of datasets {len(cfg.src_file_name)}"
+                )
+
+            datasets = []
+            for src_file, tgt_fille in zip(cfg.src_file_name, cfg.tgt_file_name):
+                if cfg.dataset_type == 'bin_memmap':
+                    dataset = BinarizedMemmapSequenceToSequenceDataset(
+                        src_dataset_prefix=src_file,
+                        tgt_dataset_prefix=tgt_fille,
+                        src_tokenizer=self.encoder_tokenizer,
+                        tgt_tokenizer=self.decoder_tokenizer,
+                        max_src_seq_length=cfg.max_seq_length,
+                        max_tgt_seq_length=cfg.max_seq_length,
+                        start_index=0,
+                        end_index=None,
+                        data_impl="mmap",
+                        skip_warmup=True,
+                    )
+                elif cfg.dataset_type == 'text_memmap':
+                    dataset = TextMemmapSequenceToSequenceDataset(
+                        src_file_name=src_file,
+                        tgt_file_name=tgt_fille,
+                        src_tokenizer=self.encoder_tokenizer,
+                        tgt_tokenizer=self.decoder_tokenizer,
+                        max_src_seq_length=cfg.max_seq_length,
+                        max_tgt_seq_length=cfg.max_seq_length,
+                    )
+                datasets.append(dataset)
+            dataset = BlendableDataset(datasets=datasets, weights=cfg.concat_sampling_probabilities)
```

Comment:
Not exactly relevant to this particular PR, but looks like BlendableDataset takes the first few samples starting from index 0 for a particular dataset. Couldn't this lead to some examples of the dataset never being used since the order is static starting from index 0?